Mathlib/CategoryTheory/Sites/Preserves.lean

```diff
@@ -0,0 +1,179 @@
+
+variable {C : Type u} [Category.{v} C] {I : C} (F : Cᵒᵖ ⥤ Type (max u v))
+
+open Limits Opposite
+
+variable (hF : (ofArrows (X := I) Empty.elim instIsEmptyEmpty.elim).IsSheafFor F)
+
+section Terminal
+
+variable (I) in
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on any
+object, then `F` takes that object to the terminal object.
+-/
+noncomputable
+def isTerminal_of_isSheafFor_empty_presieve : IsTerminal (F.obj (op I)) := by
+  refine @IsTerminal.ofUnique _ _ _ fun Y ↦ ?_
+  choose t h using hF (by tauto) (by tauto)
+  exact ⟨⟨fun _ ↦ t⟩, fun a ↦ by ext; exact h.2 _ (by tauto)⟩
+
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on the
+initial object, then `F` preserves terminal objects.
+-/
+noncomputable
+def preservesTerminalOfIsSheafForEmpty (hI : IsInitial I) : PreservesLimit (Functor.empty Cᵒᵖ) F :=
+  haveI := hI.hasInitial
+  (preservesTerminalOfIso F
+    ((F.mapIso (terminalIsoIsTerminal (terminalOpOfInitial initialIsInitial)) ≪≫
+    (F.mapIso (initialIsoIsInitial hI).symm.op) ≪≫
+    (terminalIsoIsTerminal (isTerminal_of_isSheafFor_empty_presieve I F hF)).symm)))
+
+end Terminal
+
+section Product
+
+variable (hI : IsInitial I)
+
+-- This is the data of a particular disjoint coproduct in `C`.
+variable {α : Type} (X : α → C) [HasCoproduct X] [(ofArrows X (Sigma.ι X)).hasPullbacks]
+    (hd : ∀ i j [HasInitial C], i ≠ j →
+      IsPullback (initial.to _) (initial.to _) (Sigma.ι X i) (Sigma.ι X j))
+    [∀ i, Mono (Sigma.ι X i)]
+
+/--
+The two parallel maps in the equalizer diagram for the sheaf condition corresponding to the
+inclusion maps in a disjoint coproduct are equal.
+-/
+theorem firstMap_eq_secondMap : Equalizer.Presieve.Arrows.firstMap F X (fun j ↦ Sigma.ι X j) =
+    Equalizer.Presieve.Arrows.secondMap F X (fun j ↦ Sigma.ι X j) := by
+  ext a ⟨i, j⟩
+  simp only [Equalizer.Presieve.Arrows.firstMap, Types.pi_lift_π_apply, types_comp_apply,
+    Equalizer.Presieve.Arrows.secondMap]
+  by_cases hi : i = j
+  · subst hi
+    suffices pullback.fst (f := Sigma.ι X i) (g := Sigma.ι X i) =
+      pullback.snd (f := Sigma.ι X i) (g := Sigma.ι X i) by rw [this]
+    apply Mono.right_cancellation (f := Sigma.ι X i)
+    exact pullback.condition
+  · haveI := preservesTerminalOfIsSheafForEmpty F hF hI
+    haveI := hI.hasInitial
+    let i₁ : op (pullback (Sigma.ι X i) (Sigma.ι X j)) ≅ op (⊥_ _) :=
+      ((hd i j hi).isoPullback).op
+    let i₂ : op (⊥_ C) ≅ (⊤_ Cᵒᵖ) :=
+      (terminalIsoIsTerminal (terminalOpOfInitial initialIsInitial)).symm
+    apply_fun (F.mapIso i₁ ≪≫ F.mapIso i₂ ≪≫ (PreservesTerminal.iso F)).hom using
+      injective_of_mono _
+    simp
+
+theorem piComparison_fac' {Z : C} (π : (i : α) → X i ⟶ Z) [IsIso (Sigma.desc π)] :
+    piComparison F (fun x ↦ op (X x)) =
+    F.map ((opCoproductIsoProduct X).inv ≫ (inv (Sigma.desc π)).op) ≫
+    Equalizer.Presieve.Arrows.forkMap F X π := by
```

Comment:
```suggestion
    piComparison F (fun x ↦ op (X x)) =
      F.map ((opCoproductIsoProduct X).inv ≫ (inv (Sigma.desc π)).op) ≫
        Equalizer.Presieve.Arrows.forkMap F X π := by
```
I am not sure we specifically have a code style rule for this, but I would prefer this. As you like!